text/3478-better-spread-operator.md

```diff
@@ -0,0 +1,106 @@
+- Feature Name: better-spread-operator
+- Start Date: 2023-07-23
+- RFC PR: [rust-lang/rfcs#3478](https://github.com/rust-lang/rfcs/pull/3478)
+
+# Summary
+[summary]: #summary
+
+- Allow a value to be spread into another value of a different type
+- Allow partial field spreading
+
+# Motivation
+[motivation]: #motivation
+
+We currently have the spread operator, which allows a struct to be constructed with default values provided by other instance of the struct:
+
+```rust
+struct A {
+  foo: u8,
+  bar: u8
+}
+let a = A {foo: 1, bar: 2};
+let b = A {foo: 5, ..a};
+```
+
+But this has a caveat, we can't spread another value of a different type, even if it has exactly the same fields with the same types.
+
+```rust
+struct A {
+  foo: u8,
+  bar: u8
+}
+struct B {
+  bar: u8
+}
+
+struct C {
+  foo: u8,
+  bar: u8
+}
+let b = B {bar: 2};
+let c = C {foo: 1, bar: 2};
+
+let d = A {foo: 5, ..c}; // Compiler: Sorry you can't do that, they need to have the same type
+let a = A {foo: 5, ..b}; // Compiler: Can't do that either even if all the fields are filled
+```
+
+Especially in builder patterns with lots of options, you can end up with a lot of boilerplate just on the `.finish(self)` function
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+When you have 2 instances of 2 different structs, and need to convert between one and the other easily by initializing the fields with the same values. You'd usually do this:
```

Comment:
This use case is implied to be common, but I would suspect that you're not really leveraging the type system properly, and that's why you feel this feature is useful.

If two types have a lot of the same fields in common, why? Is there some common functionality that should be factored out into a new, separate, third type? Perhaps that third type could be a field of these two parent types? Now we no longer have these common fields to worry about.

Plus, again, names of fields are arbitrary. Unlike in something like JavaScript, they are 100% removed in the compiled code and have no bearing on the actual layout of the types. Why should combining two arbitrary structs together work the same way as extending one `HashMap` from another, considering how the former is categorically different than the latter?